designs/cloudstack-multiple-endpoints.md

```diff
@@ -0,0 +1,177 @@
+
+### Tenets
+
+* ****Simple:**** simple to use, simple to understand, simple to maintain
+* ****Declarative:**** intent based system, as close to a Kubernetes native experience as possible
+
+### Goals and Objectives
+
+As a Kubernetes administrator I want to:
+
+* Perform preflight checks when creating/upgrading clusters which span across multiple failure domains
+* Create EKS Anywhere clusters which span across multiple failure domains 
+* Upgrade EKS Anywhere clusters which span across multiple failure domains
+* Delete EKS Anywhere clusters which span across multiple failure domains
+
+### Statement of Scope
+
+**In scope**
+
+* Add support for create/upgrade/delete of EKS-A clusters across multiple Cloudstack API endpoints
+* Add test environment for CI/CD e2e tests which can be used as a second Cloudstack API endpoint 
+
+**Not in scope**
+
+* 
+
+**Future scope**
+
+* Multiple network support to handle IP address exhaustion within a zone
+
+## Overview of Solution
+
+We propose to take the least invasive solution of repurposing the CloudstackDataCenterConfig to point to multiple failure domains, each of which contains the necessary
+information for interacting with a Cloudstack failure domain. The assumption is that the necessary Cloudstack resources (i.e. image, computeOffering, ISOAttachment, etc.)
+will be available on *all* the Cloudstack API endpoints. 
+
+## Solution Details
+
+### Interface changes
+Currently, the CloudstackDataCenterConfig spec contains:
+```
+// Domain contains a grouping of accounts. Domains usually contain multiple accounts that have some logical relationship to each other and a set of delegated administrators with some authority over the domain and its subdomains
+//
+Domain string `json:"domain"`
+// Zones is a list of one or more zones that are managed by a single CloudStack management endpoint.
+Zones []CloudStackZone `json:"zones"`
+// Account typically represents a customer of the service provider or a department in a large organization. Multiple users can exist in an account, and all CloudStack resources belong to an account. Accounts have users and users have credentials to operate on resources within that account. If an account name is provided, a domain must also be provided.
+Account string `json:"account,omitempty"`
+// CloudStack Management API endpoint's IP. It is added to VM's noproxy list
+ManagementApiEndpoint string `json:"managementApiEndpoint"`
+```
+
+We would instead propose to remove all the existing attributes and instead, simply include a list of FailureDomain objects, where each FailureDomain object looks like
```

Comment:
It seems (correct me if I'm missing something) that there is a direct translation between the old format and the new one

I would vote to keep the old fields and write a transformer in the defaulter code that would populate the new fields and cleanup the old ones. Up to you if a warning is printed in the cli if the old format is used as input. This should be added as a mutation webhook as well.

That keeps the api backwards compatible and reduces the friction of introducing a new format. And at the same time, it simplifies the rest of the codebase so it only has to deal with the new format.